File: src/variablearithmetic/VariableArithmetic.java
package variablearithmetic;

import lib.io.FastReader;

import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.Map;

// https://open.kattis.com/problems/variablearithmetic

public class VariableArithmetic {

    public static void main(String[] args) throws FileNotFoundException {

        StringBuilder file = new StringBuilder("src/variablearithmetic/in");
        FastReader io = new FastReader(file);
        Map<String, Integer> vars = new HashMap<>();

        while (true) {
            String line = io.nextLine();
            if (line.equals("0")) break;

            String[] tok = line.split("\\s");
            String output = processLine(tok, vars);

            if (output.equals("assignment")) continue;

            // here output is "6". And maybe more:  "6 + width + depth"
            System.out.println(output);
        }
    }

    static boolean hasAnUndefinedVariable = false;

    // every line is a combination of four things: variable, number, +, =
    static String processLine(String[] line, Map<String, Integer> vars) {
        StringBuilder sb = new StringBuilder();
        hasAnUndefinedVariable = false;
        int sum = 0;

        try {
            if (line[1].equals("=")) {
                assignVariable(line, vars);
                return "assignment";
            }
        } catch (Exception ignored) {}

        for (String s : line) {
            if (s.equals("+")) continue;

            if (isDigit(s)) {
                sum += Integer.parseInt(s);
                continue;
            }

            // It's a variable.
            if (vars.get(s) == null) {
                sb.append(s).append(" + ");
                hasAnUndefinedVariable = true;
            } else {
                sum += vars.get(s);
            }
        }

        if (sum == 0) {
            if (hasAnUndefinedVariable) {
                sb.setLength(sb.length() - 3);
                return sb.toString();
            }
            return String.valueOf(sum);
        }

        if (!isDigit(sb.toString()) && !sb.toString().isEmpty()) {
            sb.setLength(sb.length() - 3);
            return sum + " + " + sb.toString();
        } else {
            return String.valueOf(sum);
        }
    }

    static void assignVariable(String[] s, Map<String, Integer> vars) {
        vars.put(s[0], Integer.parseInt(s[2]));
    }

    static boolean isDigit(String s) {
        try {
            Integer.parseInt(s);
            return true;
        } catch (Exception e) {
            return false;
        }
    }
}

File: src/variablearithmetic/variablearithmetic.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// https://open.kattis.com/problems/variablearithmetic

func main() {
	input := os.Getenv("GOPATH") + filepath.Join("/src/code/variablearithmetic/in")
	file, _ := os.Open(input)
	io, _ := ioutil.ReadAll(file)

	allLines := strings.Split(string(io), "\n")
	allLines = allLines[:len(allLines) - 2]				// remove linefeed and zero from tail

	var vars  = make(map[string]int)
	var sb strings.Builder
	var result []string
	var sum int64
	var hasNumbers bool

	for _, lines := range allLines {
		line := strings.Fields(lines)
		result = result[:0]
		sum = 0
		hasNumbers = false

		// if assignment (length = 3)
		if len(line) > 1 {
			if line[1] == "=" {
				num, _ := strconv.Atoi(line[2])
				vars[line[0]] = num
				continue
			}
		}

		// each line is a combo of 4 things: word, number, +, =
		// go thru line, and if digit sum it. Else check map. And trim concat words, trim
		for i := 0; i < len(line); i += 2 {
			word := line[i]
			if x, err := strconv.ParseInt(word, 0, 64); err == nil {
				sum += x
				hasNumbers = true
			} else {
				// Is a variable, check if it exists
				if val, ok := vars[word]; ok {
					sum += int64(val)
				} else {
					result = append(result, word)
				}
			}
		}

		// process the line
		num := strconv.Itoa(int(sum))
		if len(result) == 0 {
			sb.WriteString(num + "\n")
			continue
		}

		if hasNumbers {
			sb.WriteString(num + " + ")
		}
		for i := 0; i < len(result) - 1; i++ {
			sb.WriteString(result[i] + " + ")
		}
		sb.WriteString(result[len(result) - 1] + "\n")

	}
	fmt.Print(sb.String())
}
